cirq/contrib/acquaintance/strategy.py

```diff
@@ -0,0 +1,135 @@
+    from cirq.ops import Gate
+
+STRATEGY_GATE = Union[AcquaintanceOpportunityGate, PermutationGate]
+
+class AcquaintanceStrategy(Circuit):
+    """An acquaintance strategy is a special type of circuit that contains only
+    gates of two types: AcquaintanceOpportunityGate and PermutationGate. It
+    serves as a skeleton for compiling around geometric constraints.
+    """
+
+    gate_types = (AcquaintanceOpportunityGate, PermutationGate)
+
+    def __init__(self, circuit: Circuit=None) -> None:
+        if circuit is None:
+            circuit = Circuit()
+        for moment in circuit:
+            for op in moment.operations:
+                if not isinstance(op.gate, self.gate_types):
+                    raise ValueError('An acquaintance strategy can only'
+                        'contain gates of types {}'.format(self.gate_types))
+        super().__init__(circuit._moments)
+
+    def rectify(self, acquaint_first: bool=True):
+        """Splits moments so that they contain either only acquaintance gates
+        or only permutation gates."""
+        last_gate_type = self.gate_types[int(not acquaint_first)]
+        rectified_moments = []
+        for moment in self:
+            gate_type_to_ops = defaultdict(list
+                    ) # type: Dict[Type[STRATEGY_GATE], List[GateOperation]]
+            for op in moment.operations:
+                gate_type_to_ops[type(op.gate)].append(op)
+            if len(gate_type_to_ops) == 1:
+                rectified_moments.append(moment)
+                continue
+            for gate_type in sorted(gate_type_to_ops,
+                                    key=partial(ne, last_gate_type)):
+                rectified_moments.append(Moment(gate_type_to_ops[gate_type]))
+                last_gate_type = gate_type
+        self._moments = rectified_moments
+
+    def nest(self, qubit_order: Sequence[QubitId], acquaintance_size: int=0
+            ) -> bool:
+        """Replace every moment containing acquaintance gates (after
+        rectification) with a generalized swap network, with the partition
+        given by the acquaintance gates in that moment (and singletons for the
+                free qubits). Accounts for reversing effect of swap networks."""
+        self.rectify()
+        reflected = False
+        reverse_map = {q: r for q, r in zip(qubit_order, reversed(qubit_order))}
+        for moment_index, moment in enumerate(self):
+            if reflected:
+                moment = moment.transform_qubits(reverse_map.__getitem__)
+            if all(isinstance(op.gate, AcquaintanceOpportunityGate)
+                    for op in moment.operations):
+                swap_network_gate = SwapNetworkGate.from_operations(
+                        qubit_order, moment.operations, acquaintance_size)
+                swap_network_op = swap_network_gate(*qubit_order)
+                moment = Moment([swap_network_op])
+                reflected = not reflected
```

Comment:
Is it actually guaranteed that this was a reflection? What if there's only one operation in the moment?

Reply:
`swap_network_gate` creates singleton parts for any qubit not acted on, so yes, this is guaranteed to be a reflection (over `nest`'s `qubit_order` argument).